python/src/iceberg/expression/literals.py

```diff
@@ -0,0 +1,742 @@
+#  Licensed under the Apache License, Version 2.0 (the "License");
+#  you may not use this file except in compliance with the License.
+#  You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+#  Unless required by applicable law or agreed to in writing, software
+#  distributed under the License is distributed on an "AS IS" BASIS,
+#  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#  See the License for the specific language governing permissions and
+#  limitations under the License.
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+
+import datetime
+import sys
+import uuid
+from decimal import ROUND_HALF_UP, Decimal
+from functools import singledispatch
+
+if sys.version_info >= (3, 8):
+    from functools import singledispatchmethod
+else:
+    from singledispatch import singledispatchmethod
+
+import pytz
+
+from iceberg.types import (
+    BinaryType,
+    BooleanType,
+    DateType,
+    DecimalType,
+    DoubleType,
+    FixedType,
+    FloatType,
+    IntegerType,
+    LongType,
+    Singleton,
+    StringType,
+    TimestampType,
+    TimestamptzType,
+    TimeType,
+    UUIDType,
+)
+
+JAVA_MAX_INT = 2147483647
+JAVA_MIN_INT = -2147483648
+JAVA_MAX_FLOAT = 3.4028235e38
+JAVA_MIN_FLOAT = -3.4028235e38
+EPOCH = datetime.datetime.utcfromtimestamp(0)
+EPOCH_DAY = EPOCH.date()
+
+"""
+Iceberg literal is wrapper class used in expressions, which return unbound predicates
+It's being organized as below
+Literal
+|-- AboveMax
+|-- BelowMin
+|-- BaseLiteral
+    |-- StringLiteral
+    |-- FixedLiteral
+    |-- BinaryLiteral
+    |-- ComparableLiteral
+        |-- BooleanLiteral
+        |-- IntegerLiteral
+        |-- LongLiteral
+        |-- FloatLiteral
+        |-- DoubleLiteral
+        |-- DateLiteral
+        |-- TimeLiteral
+        |-- TimestampLiteral
+        |-- DecimalLiteral
+        |-- UUIDLiteral
+"""
+
+
+class Literal:
+    def to(self, type_var):
+        raise NotImplementedError()
+
+    def to_byte_buffer(self):
+        raise NotImplementedError()
+
+
+@singledispatch
+def of(value):
+    """
+    A generic Literal factory to construct an iceberg Literal based on python primitive data type
+    using dynamic overloading
+
+    Args:
+        value(python primitive type): the value to be associated with literal
+
+    Example:
+        import iceberg.expressions.literals
+        >>> iceberg.expressions.literals.of(1)
+        IntegerLiteral(1)
+    """
+    raise TypeError(f"Unimplemented Type Literal for value: {value}")
+
+
+@of.register(bool)
+def _of(value):
```

Comment:
Is it feasible to put type annotation for `value`?

Reply:
there's type annotation on the decorator above, do you think if it's sufficient? or I can add for `value` if you prefer

Reply:
`@singledispatch` is actually clever enough to look at the typehint, so you could actually do this:
```py
@of
def _(value: bool):
    ...
```
The one downside is that if you want to dispatch one function for multiple types, you can't stack them the way you can with the decorator that calls register and unfortunately it doesn't recognize the `typing.Union` type hint where you specify multiple types. That's not a concern here though since it's a 1:1 mapping between types and logic.

Reply:
yep, I ended up with moving type annotation to parameter instead of decorator since they are 1:1 mapping as of now